pkg/ebpf/compiler/compiler.cpp

```diff
@@ -0,0 +1,273 @@
+        LLVMInitializeBPFTargetInfo();
+        LLVMInitializeBPFAsmPrinter();
+        LLVMInitializeBPFAsmParser();
+
+        ClangCompiler::llvmInitialized = true;
+    }
+
+    theDriver = std::make_unique<clang::driver::Driver>(
+        name, getArch(), *diagnosticsEngine,
+        llvm::vfs::getRealFileSystem()
+    );
+
+    std::string arch = "bpf";
+    std::string Error;
+
+    theTarget = llvm::TargetRegistry::lookupTarget(theTriple.getTriple(), Error);
+    if (!theTarget) {
+        errString = "could not lookup target";
+        return;
+    }
+
+    llvm::TargetOptions targetOptions;
+    auto RM = llvm::Optional<llvm::Reloc::Model>();
+    targetMachine = std::unique_ptr<llvm::TargetMachine>(theTarget->createTargetMachine(
+        theTriple.getTriple(), "generic", "", targetOptions, RM, llvm::None, llvm::CodeGenOpt::Aggressive));
+
+    if (!targetMachine) {
+        errString = "could not allocate target machine";
+        return;
+    }
+
+    if (remapped_files_.empty()) {
+        for (auto f : MappedFiles::files) {
+            remapped_files_[f.first] = llvm::MemoryBuffer::getMemBuffer(f.second);
+        }
+    }
+}
+
+std::unique_ptr<clang::CompilerInvocation> ClangCompiler::buildCompilation(
+    const char *inputFile,
+    const char *outputFile,
+    const std::vector<const char*> &extraCflags,
+    bool verbose)
+{
+    std::vector<const char*> cflags;
+    for (auto it = defaultCflags.begin(); it != defaultCflags.end(); it++)
+        cflags.push_back(*it);
+    for (auto it = extraCflags.begin(); it != extraCflags.end(); it++)
+        cflags.push_back(*it);
+
+    if (verbose) {
+        cflags.push_back("-v");
+    }
+
+    cflags.push_back("-c");
+    cflags.push_back(inputFile);
+
+    if (outputFile) {
+        cflags.push_back("-o");
+        cflags.push_back(outputFile);
+    }
+
+    // Build
+    std::unique_ptr<clang::driver::Compilation> compilation(theDriver->BuildCompilation(cflags));
+
+    // expect exactly 1 job, otherwise error
+    const clang::driver::JobList &jobs = compilation->getJobs();
+    if (jobs.size() != 1 || !clang::isa<clang::driver::Command>(*jobs.begin())) {
+        clang::SmallString<256> msg;
+        llvm::raw_svector_ostream os(msg);
+        jobs.Print(os, "; ", true);
+        diagnosticsEngine->Report(clang::diag::err_fe_expected_compiler_job) << os.str();
+        return nullptr;
+    }
+
+    const clang::driver::Command &cmd = clang::cast<clang::driver::Command>(*jobs.begin());
+    if (llvm::StringRef(cmd.getCreator().getName()) != "clang") {
+        diagnosticsEngine->Report(clang::diag::err_fe_expected_clang_command);
+        return nullptr;
+    }
+
+    if (compilation->containsError()) {
+        return nullptr;
+    }
+
+    if (verbose) {
+        llvm::errs() << "clang invocation:\n";
+        jobs.Print(llvm::errs(), "\n", true);
+        llvm::errs() << "\n";
+    }
+
+    std::unique_ptr<clang::CompilerInvocation> invocation(new clang::CompilerInvocation);
```

Comment:
Can be replaced by `auto invocation = std::make_unique<clang::CompilerInvocation>()`